src/libcore/ptr.rs

```diff
@@ -8,11 +8,68 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// FIXME: talk about offset, copy_memory, copy_nonoverlapping_memory
-
-//! Raw, unsafe pointers, `*const T`, and `*mut T`.
+//! Manually manage memory through raw pointers.
 //!
 //! *[See also the pointer primitive types](../../std/primitive.pointer.html).*
+//!
+//! # Safety
+//!
+//! Many functions in this module take raw pointers as arguments and read from
+//! or write to them. For this to be safe, these pointers must be *valid*.
+//! Whether a pointer is valid depends on the operation it is used for
+//! (read or write), and the extent of the memory that is accessed (i.e.,
+//! how many bytes are read/written). Most functions use `*mut T` and `*const T`
+//! to access only a single value, in which case the documentation omits the size
+//! and implicitly assumes it to be `size_of::<T>()` bytes.
+//!
+//! The precise rules for validity are not determined yet.  The guarantees that are
+//! provided at this point are very minimal:
+//!
+//! * A [null] pointer is *never* valid, not even for accesses of [size zero][zst].
```

Comment:
This isn't always true in bare-metal scenarios (or even some obscure OSes probably). I'm fine with rust defining it like this though, just want to make sure it's a conscious decision, users will just have to use inline assembly to read/write possibly null pointers.

Reply:
Do you know of any target in which this is the case that's minimally in use? 

In C null pointer constant expressions do not have to refer to the address at `0_usize`, but AFAIK modern C compilers do not support this because there are no available targets that require it.

Reply:
@parched You can have a pointer with the same object representation as a null pointer, but which is not a null pointer; it's just really hard to construct. I'd look at C's guarantees around this.

Reply:
@gnzlbg x86 OS code in real mode, with the interrupt vector table at address 0. (That *is* still in real use, especially for things like bootloaders.)

Reply:
@joshtriplett sure but do people write those using C ? Because AFAIK you have to use non-standard C compilers (e.g. openwatcom can target i286) to target those in C + extensions _without_ assembly.

Reply:
GCC has `-fno-delete-null-pointer-checks` for this I believe. See also LLVM discussion [here](http://lists.llvm.org/pipermail/llvm-dev/2018-April/122632.html)

Reply:
@gnzlbg You can target 32-bit real mode with GCC, and use a bit of `.code16` assembly to switch to 32-bit big real mode.

Reply:
@joshtriplett you can target 16-bit x86 with clang, but you don't have near/far/huge pointer support, so have to target a single segment, and use near pointers everywhere.

In any case, this discussion belongs in the unsafe code guidelines where this issue has already been raised.

Reply:
AFAIK LLVM makes the assumption that NULL is never in-bounds of any allocation (in address space 0, anyway). We set all sorts of `nonnull` attributes as well. I see no way to support actually writing to NULL while using LLVM as backend, currently.

Also, lacking a proper definition of the abstract representation of a pointer in LLVM, I'd consider all pointers to be NULL pointers whose in-memory representation ("object representation", as @ubsan would call it) consists of only 0.

Reply:
Moved to https://github.com/rust-rfcs/unsafe-code-guidelines/issues/29.